CERT-Polska/mwdb-core
# Rich-text attribute rendering

**Your checklist for this pull request**

**What is the current behaviour?**
Attributes support JSON objects and simple URLs. Any rich rendering requires plugin development.

**What is the new behaviour?**

PoC of rich-text attribute rendering templates as a part of https://github.com/CERT-Polska/mwdb-core/issues/413.

![image](https://user-images.githubusercontent.com/8720367/167902103-46294b57-7fdc-4203-90c5-d5a951511f6f.png)

Target is to implement features that enable us to produce similar reports like Details section in VirusTotal (e.g. https://www.virustotal.com/gui/file/32fae9922417d6405bf60144d819d5e02b44060fa8f07e5e71c824725f44307f/details) based on data contained in attribute objects. Attribute templates will be defined by MWDB administrator, but we can come with some built-in handy templates.

Template language is based on [Mustache](https://github.com/janl/mustache.js#variables) which is then rendered to Markdown. Then we use [Marked.js](https://github.com/markedjs/marked) to render it into React components. Both libraries are a bit modified in this PR because their target language was HTML and it's not that easy to change without some heavy overrides.

**Things to do:**

1. Changes in model:
   - Additional field in `AttributeDefinition` called `rich_template` to store template for attribute key.
   - URL templates should be transformed into Rich templates by migration
     `http://xxx.cert.pl/case/$value` => `[{{value}}](http://xxx.cert.pl/case/{{value}})`
     `url_template` can be removed then or kept for backwards compatibility.
2. Place interactive template editor in attribute key definition view. It can be similar to the one provided in PoC PR.
3. Add rich rendering into object view. We should be able to switch between raw and rich representation for each attribute.

**Possible future extensions**
- Collapsible trees e.g. `*` may be collapsible items and `-` regular ones
  ```
  * KERNEL32.dll
     - 1111
  ```
- Optional value transformation into interactive link that redirect to search for specific attribute value
- Support for widgets to place ready-made React component instead of Markdown primitive (for buttons, graphs etc.). These widgets should be extensible for plugins.

**Test plan**

- Make sure we don't incorporate any XSS with this feature. It should not be a problem if we operate only on React objects and not apply direct DOM transformations.
- Check if escaping works correctly. HTML escaping is hardwired in Marked, but we can fix it by Tokenizer overrides.
- Check if values injected by Mustache are escaped correctly and doesn't interfere with Markdown.
- Check if all Markdown element types render correctly. Some of them are intentionally ignored:
    - We don't want to have images rendered, doesn't sound safe with user-controlled input
    - Raw HTML, headings and other weird things should be rendered as regular text.
